### PR TITLE
fix(use-output-allocator): report graph-output rank for reshaped-view outputs

### PR DESCRIPTION
**Pull request by anilmartha**
## Summary

`UseOutputAllocator` rewrote a graph-output `memref.alloc` into `hip.alloc_output` using the **producer alloc's type**. When the returned value is a reshaping view of that alloc (e.g. a rank-2 Gemm result `expand_shape`'d to rank-3 with no trailing op — DETR's `class_labels_classifier` `logits` head), the EP registered the alloc's rank with ORT instead of the declared graph-output rank, so ORT rejected the output:

```
Invalid rank for output: logits Got: 2 Expected: 3
```

Outputs with a trailing op (e.g. `pred_boxes`' `sigmoid`, whose alloc is already at the output rank) were unaffected — which is why only `logits` failed.

## Fix

- Allocate `hip.alloc_output` at the `func.return` operand type (the graph-output shape).
- When it differs from the producer alloc's type, rebuild the **inverse** of the view chain (`cast` / `expand_shape` / `collapse_shape`, static output) so the producer keeps writing into the same EP-owned buffer viewed as its original type — no extra copy.
- Fast path (return type == alloc type) is unchanged; `subview` or dynamic-shaped outputs fall back to the legacy behavior, so no existing case regresses.

Applies to both the `LLVM_IR` and `NATIVE` artifact paths.

## Test plan

- [x] `facebook/detr-resnet-50` (fp16): `logits` now returns `{1,100,92}`; `onnxruntime_perf_test` runs clean.
- [ ] Re-run existing lit/numeric suites for regressions (esp. conv/reshape output models).
- [ ] Spot-check a model with a rank-4 conv output collapsed before return.

### CODE REVIEW

**Comment by amd-mingw** [comment]
https://github.com/ROCm/hip-ep/pull/533 is fixing on it. cc @amd-genmingz
